homeassistant/components/esphome/bluetooth/async_interrupt.py

```diff
@@ -0,0 +1,106 @@
+"""Interrupt context manager for asyncio.
+
+This module provides a context manager that can be used to interrupt
+a block of code when a future is done.
+
+The purpose is to raise as soon as possible to avoid any race conditions.
+
+This is based loosely on async_timeout by Andrew Svetlov and cpython asyncio.timeout
+"""
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Callable
+from functools import partial
+from types import TracebackType
+from typing import TYPE_CHECKING, Any, final
+
+__version__ = "1.0.0"
+
+__all__ = ("interrupt",)
+
+
```

Comment:
This will live at https://github.com/bdraco/async_interrupt after I finish testing